blockchain/utxocache.go

```diff
@@ -0,0 +1,793 @@
+	b.chainLock.RUnlock()
+	return view, err
+}
+
+// Commit commits all the entries in the view to the cache.
+//
+// This function is safe for concurrent access.
+func (s *utxoCache) Commit(view *UtxoViewpoint) error {
+	for outpoint, entry := range view.Entries() {
+		// No need to update the database if the entry was not modified or fresh.
+		if entry == nil || (!entry.isModified() && !entry.isFresh()) {
+			continue
+		}
+
+		// We can't use the view entry directly because it can be modified
+		// lateron.
+		ourEntry := s.cachedEntries[outpoint]
+		if ourEntry == nil {
+			ourEntry = entry.Clone()
+		}
+
+		// Remove the utxo entry if it is spent.
+		if entry.IsSpent() {
+			if err := s.spendEntry(outpoint, ourEntry); err != nil {
+				return err
+			}
+			continue
+		}
+
+		// Store the entry we don't know.
+		if err := s.addEntry(outpoint, ourEntry, false); err != nil {
+			return err
+		}
+	}
+
+	view.prune()
+	return nil
+}
+
+// flush flushes the UTXO state to the database.
+//
+// This method should be called with the state lock held.
+func (s *utxoCache) flush(bestState *BestState) error {
+	// If we performed a flush in the current best state, we have nothing to do.
+	if bestState.Hash == s.lastFlushHash {
+		return nil
+	}
+
+	// Add one to round up the integer division.
+	totalMiB := s.totalMemoryUsage()/(1024*1024) + 1
+	log.Infof("Flushing UTXO cache of ~%v MiB to disk. For large sizes, "+
+		"this can take up to several minutes...", totalMiB)
+
+	// First update the database to indicate that a utxo state flush is started.
+	// This allows us to recover when the node shuts down in the middle of this
+	// method.
+	err := s.db.Update(func(dbTx database.Tx) error {
+		return dbPutUtxoStateConsistency(dbTx, ucsFlushOngoing, &s.lastFlushHash)
+	})
+	if err != nil {
+		return err
+	}
+
+	// Store all entries in batches.
+	flushBatch := func(dbTx database.Tx) error {
+		nbBatchEntries := 0
+		for outpoint, entry := range s.cachedEntries {
+			// Nil entries or unmodified entries can just be pruned.
+			// They don't count for the batch size.
+			if entry == nil || !entry.isModified() {
+				s.totalEntryMemory -= entry.memoryUsage()
+				delete(s.cachedEntries, outpoint)
+				continue
+			}
+
+			if entry.IsSpent() {
+				if err := dbDeleteUtxoEntry(dbTx, outpoint); err != nil {
+					return err
+				}
+			} else {
+				if err := dbPutUtxoEntry(dbTx, outpoint, entry); err != nil {
+					return err
+				}
+			}
+			nbBatchEntries++
+
+			s.totalEntryMemory -= entry.memoryUsage()
+			delete(s.cachedEntries, outpoint)
+
+			// End this batch when the maximum number of entries per batch has
+			// been reached.
+			if nbBatchEntries >= utxoBatchSizeEntries {
+				break
+			}
+		}
+		return nil
+	}
+	for len(s.cachedEntries) > 0 {
+		log.Tracef("Flushing %d more entries...", len(s.cachedEntries))
+		err := s.db.Update(func(dbTx database.Tx) error {
+			return flushBatch(dbTx)
+		})
+		if err != nil {
+			return err
```

Comment:
roll back to latest stable state?

Reply:
That error means a disk error of some sort. Likely that all further IO operations would also error. Besides, there is logic to recover on startup when shut down in the middle of a flush.